### ISSUE TITLE
Build error: grpc.SupportPackageIsVersion3

### ISSUE THREAD

**Issue by buchanae** [bmeg/task-execution-server#24]
I cloned the repo and followed the instructions in the README. `make` fails with an error.

```
$> make depends
go get -u github.com/golang/protobuf/protoc-gen-go
go get -u github.com/grpc-ecosystem/grpc-gateway/protoc-gen-swagger
go get -u github.com/grpc-ecosystem/grpc-gateway/protoc-gen-grpc-gateway
go get -d tes-server/
go get -d tes-worker/

$> make
go install tes-server
# tes/ga4gh
src/tes/ga4gh/task_execution.pb.go:432: undefined: grpc.SupportPackageIsVersion3
make: *** [server] Error 2
```

**Comment by jeenalee** [comment]
Hi @buchanae, thank you for opening this issue.
I wanted to double check: is this with `develop` branch?

**Comment by buchanae** [comment]
I pulled the `master` branch first. I also just tried `develop` and had the same error.

I suspect the problem is that this project isn't installing specific versions of dependencies (just he "latest" version) and gRPC has been updated to a newer, backwards-incompatible version.

I'm new to `go`, but seems like this is the behavior for package management out of the box, as described here: http://jbeckwith.com/2015/05/29/dependency-management-go/

So, maybe the next best step is to use `godep`?

**Comment by jeenalee** [comment]
I gave the link a quick skim, and I think that's the reason too.

I deleted the built dependencies, did `make depends` and `make`, and I saw the same error you described above.

**Comment by buchanae** [comment]
We should also determine the most recent recommendations for go deps. I bet things have changed since that article was written. Sounds like go 1.7 has a built in vendoring approach.

**Comment by buchanae** [comment]
https://golang.org/cmd/go/#hdr-Vendor_Directories

**Comment by jeenalee** [comment]
If we were to go with using `vendor` approach, does that mean we will ship TES with the packages locally built? 

What approach is better in your opinion?

**Comment by buchanae** [comment]
@jeenalee not sure what you mean by "with the packages locally built". We can still ship TES binary releases, if that's what you mean.

As I understand, "vendoring/vendored" means committing the dependency code into the TES repository.

The question is, what's the best way to manage the code in the `vendor` directory. The Go community is still exploring this and no winner has emerged. I'll try to summarize what I've learned so far. The approaches are:

1. Commit the dependencies into the TES codebase. 
---

Pros
* the dependencies are obvious, exact, and readily available
* less chance of issues downloading dependencies (e.g. if github is down) because they're already in the repo.
* easier to set up a build server, because no need to manage dependencies.

Cons
* codebase gets bigger and heavier
* difficult to update dependencies?
* no way to tell which version of a dependency you have, unless you write them down

This is the approach initially recommended by the Go core team, although it's been argued that Google's approach doesn't work for everybody.

2. Use git submodules
---

Pros
* git submodule tools make it easier to update and track dependency code
* `go get` supports git submodules, so this still fits in with the core Go approach

Cons
* limited to git
* some say git submodule tools are still clumsy

The [manul](https://github.com/kovetskiy/manul) tool takes this approach.
This [vendetta](https://github.com/dpw/vendetta) tool does too.

3. Use a dependency list file (a la package.json or requirements.txt) and a package manager
---

Pros
* Familiar: lots of languages work this way (node.js, python, ruby, etc)
* Dependencies and versions are written down in a file

Cons
* No standard tool or file format has been decided.
* Cannot `go get` the package because it depends on a separate package management tool.

4. Use a dependency file AND vendor the dependencies
---

So, use a tool (such as glide) to list and download dependencies during development, but commit them to the repo so you still get the benefits of approaches 1 and 2.


I tried [glide](https://glide.sh/) yesterday. I thought it was really nice, but I only have about 5 minutes of experience with it.

I also tried [godep](https://github.com/tools/godep) but I couldn't get it working and/or didn't understand how to use it. Also, it wraps the core `go` commands (e.g. `godep go install`) which is unfortunate.

I'm not sure which is best. I'm sort of leaning toward git submodules; they're simple and should solve most, if not all, of the needs we have, without requiring an external, non-standard tool.

**Comment by jeenalee** [comment]
@buchanae Hey Alex! Are you working on this issue?

**Comment by buchanae** [comment]
I haven't worked on it since my last comment. We (you, me, and Kyle at least) should get together and chat and make a decision about our direction with package management.

**Comment by kellrott** [comment]
Update of code to new GRPC code in https://github.com/bmeg/task-execution-server/pull/26
Note: this doesn't address the dependency management

**Comment by buchanae** [comment]
@jeenalee Did you make any progress on this?

**Comment by simonovic86** [comment]
Hi. I have the same problem here. Btw, I'm also new to Go. Is there any workaround for this?

**Comment by jeenalee** [comment]
@simonovic86 Apologies for the late response. We had this build issue because we had updated a protobuf file but hadn't updated the protobuf generated go code. Once we re-built the `proto.go` files, the issue went away.

Though updating protobuf generated code fixed this issue for us, this build issue could also happen because of package dependencies. If a package your software uses was updated and the newer version doesn't comply with your software, this build issue could happen. There are a few ways to manage Go dependencies, and it appears there isn't a clear winner. @buchanae [compared some of them previously](https://github.com/bmeg/task-execution-server/issues/24#issuecomment-259201473). We ended up using git submodules. Hope this helps!
